### DISCUSSION TITLE
get_target_property of INCLUDE_DIRECTORIES on CGAL::CGAL is not set

### DISCUSSION THREAD

**Discussion by VincentRouvreau** [CGAL/cgal#8148] [General]
## Issue Details

It would be nice if `get_target_property(CGAL_TARGET_INCLUDE_DIRECTORIES CGAL::CGAL INCLUDE_DIRECTORIES)` was returning the CGAL include directory.

## Source Code

Here is a CMake file to reproduce the behaviour:
```cmake
cmake_minimum_required(VERSION 3.8)
project(minimal_cgal_dependency)

find_package(CGAL QUIET)

get_target_property(CGAL_TARGET_INTERFACE_INCLUDE_DIRECTORIES CGAL::CGAL INTERFACE_INCLUDE_DIRECTORIES)
message("*** CGAL_TARGET_INTERFACE_INCLUDE_DIRECTORIES = ${CGAL_TARGET_INTERFACE_INCLUDE_DIRECTORIES}")
# This seems OK, but not convenient
# *** CGAL_TARGET_INTERFACE_INCLUDE_DIRECTORIES = /miniconda3/envs/custom-cgal/include;/miniconda3/envs/custom-cgal/include;/miniconda3/envs/custom-cgal/include;$<BUILD_INTERFACE:/workspace/contrib/CGAL-5.5.3/include>;$<INSTALL_INTERFACE:include>

get_target_property(CGAL_TARGET_INCLUDE_DIRECTORIES CGAL::CGAL INCLUDE_DIRECTORIES)
message("*** CGAL_TARGET_INCLUDE_DIRECTORIES = ${CGAL_TARGET_INCLUDE_DIRECTORIES}")
# This behaviour would be convenient
# *** CGAL_TARGET_INCLUDE_DIRECTORIES = CGAL_TARGET_INCLUDE_DIRECTORIES-NOTFOUND

message("*** CGAL_INCLUDE_DIRS = ${CGAL_INCLUDE_DIRS}")
# The 'old' behaviour is still ok
# *** CGAL_INCLUDE_DIRS = /workspace/contrib/CGAL-5.5.3/include

...
```

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): Linux 64 bits
* Compiler: g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
* Release or debug mode: Release mode
* Specific flags used (if any):
* CGAL version: Same behaviour with CGAL 5.1.5, 5.5.3 and 6.0-I-207 (master from some days ago)
* Boost version: 1.84.0
* Other libraries versions if used (Eigen, TBB, etc.):
cmake                   3.29.0               hcfe8598_0    conda-forge
eigen                     3.4.0                h00ab1b0_0    conda-forge
gmp                       6.3.0                h59595ed_1    conda-forge
mpfr                      4.2.1                h9458935_0    conda-forge

**Comment by lrineau** [comment]
I do not get it. Target properties like `INCLUDE_DIRECTORIES` or `INTERFACE_INCLUDE_DIRECTORIES` are not meant to be consumed by users of the target. They are set by CMake.

`CGAL::CGAL` should actually have `INTERFACE_INCLUDE_DIRECTORIES` empty, in favor of `INTERFACE_SYSTEM_INCLUDE_DIRECTORIES`.

`${CGAL_INCLUDE_DIRS}` is the contract we have with CGAL users, and should stay stable over versions of CGAL. Why does it not fit your needs? What are you trying to do?

**Reply by VincentRouvreau** [reply]
Sorry if I am out of topic, I am far to be a CMake expert...
But if I read [this stackoverflow](https://stackoverflow.com/a/58767906/5092461), it seems to be the _modern way_ to get include/lib directory for a target.

And yes, for sure, you can keep `${CGAL_INCLUDE_DIRS}` for backward compatibility, I do not have any issue with this (and I will keep using it).

**Reply by lrineau** [reply]
In modern CMake, you usually link with imported targets like `CGAL::CGAL`, and do not deal with include dirs, libraries, definitions, compiler options...

Vincent, if you explain why you need the include directory of CGAL, I might help you with that.

**Reply by VincentRouvreau** [reply]
Yes, for sure. As you may know, gudhi is using some nice cgal features. So for the c++ part, it is quite easy with cmake:
```cmake
...
find_package(Boost REQUIRED)
find_package(CGAL)

if (CGAL_FOUND)
  add_executable ( some_gudhi_program some_gudhi_program.cpp )
  target_link_libraries(some_gudhi_program  CGAL::CGAL)
endif()
```

(I tried also [scikit-build](https://github.com/VincentRouvreau/hands_on_scikit_build) which seems a good and modern candidate, but I was having issues regarding third parties installation with msvc)
But when we do the python bindings, we need to construct a `setup.py` file with all the user setup. To do so (maybe not the best idea, but this is the design we have chosen), we ask cmake to build this `setup.py` file.

This is why we need cmake to know from CGAL installation if it is a _header only_ version or if we have to link with some dynamic libraries (here we need its path too), and where are the CGAL header files. This gives us a `setup.py` file that looks like (simplified version):

```python
from setuptools import setup, Extension, find_packages
from Cython.Build import cythonize

cython_modules = ['simplex_tree', 'alpha_complex', ]

source_dir='/workspace/gudhi/gudhi-devel/src/python/gudhi/'
extra_compile_args=['-DBOOST_RESULT_OF_USE_DECLTYPE', '-DBOOST_ALL_NO_LIB', '-DBOOST_SYSTEM_NO_DEPRECATED', '-std=c++17', '-frounding-math', '-DCGAL_EIGEN3_ENABLED', '-DCGAL_HEADER_ONLY', '-DCGAL_USE_GMP', '-DCGAL_USE_GMPXX', '-DCGAL_USE_MPFR', '-DGUDHI_USE_TBB', ]
extra_link_args=['-s', ]
libraries=['gmp', 'gmpxx', 'mpfr', 'tbb', 'tbbmalloc', ]
library_dirs=['/miniconda3/envs/gudhi-devel/lib', ]
include_dirs = ['/workspace/contrib/CGAL-5.1.5/include', '/miniconda3/envs/gudhi-devel/include', '/workspace/gudhi/gudhi-devel/src/python/gudhi/', '/workspace/gudhi/gudhi-devel/src/Alpha_complex/include', '/workspace/gudhi/gudhi-devel/src/Simplex_tree/include', ]
runtime_library_dirs=['/miniconda3/envs/gudhi-devel/lib', ]

# Create ext_modules list from module list
ext_modules = []
for module in cython_modules:
    ext_modules.append(Extension(
        'gudhi.' + module,
        sources = [source_dir + module + '.pyx',],
        language = 'c++',
        extra_compile_args=extra_compile_args,
        extra_link_args=extra_link_args,
        libraries=libraries,
        library_dirs=library_dirs,
        include_dirs=include_dirs,
        runtime_library_dirs=runtime_library_dirs,))

ext_modules = cythonize(ext_modules, compiler_directives={'language_level': '3'})

setup(
    name = 'gudhi',
    packages=find_packages(),
    author='GUDHI Editorial Board <https://gudhi.inria.fr/contact/>',
    version='3.10.0a1',
    ext_modules = ext_modules,
    python_requires='>=3.5.0',
    install_requires = ['numpy >= 1.15.0',],
    package_data={"": ["*.dll"], },
)
```

**Comment by VincentRouvreau** [comment]
Just a reminder that my initial question was to know if it is _"normal"_ that the target property `INCLUDE_DIRECTORIES` of `CGAL::CGAL` was not set :sweat_smile:

**Comment by VincentRouvreau** [comment]
> `CGAL::CGAL` should actually have `INTERFACE_INCLUDE_DIRECTORIES` empty, in favor of `INTERFACE_SYSTEM_INCLUDE_DIRECTORIES`.

:thinking: Interesting ! Maybe I was not asking for the correct target property... Here is what I have for CGAL in comparison with TBB and Eigen3, for instance:
```cmake
get_target_property(EIGEN3_INTERFACE_SYSTEM_INCLUDE_DIRECTORIES Eigen3::Eigen INTERFACE_SYSTEM_INCLUDE_DIRECTORIES)
# EIGEN3_INTERFACE_SYSTEM_INCLUDE_DIRECTORIES=EIGEN3_INTERFACE_SYSTEM_INCLUDE_DIRECTORIES-NOTFOUND
get_target_property(EIGEN3_INTERFACE_INCLUDE_DIRECTORIES Eigen3::Eigen INTERFACE_INCLUDE_DIRECTORIES)
# EIGEN3_INTERFACE_INCLUDE_DIRECTORIES=/miniconda3/envs/custom-cgal/include/eigen3

get_target_property(TBB_INTERFACE_SYSTEM_INCLUDE_DIRECTORIES TBB::tbb INTERFACE_SYSTEM_INCLUDE_DIRECTORIES)
# TBB_INTERFACE_SYSTEM_INCLUDE_DIRECTORIES=TBB_INTERFACE_SYSTEM_INCLUDE_DIRECTORIES-NOTFOUND
get_target_property(TBB_INTERFACE_INCLUDE_DIRECTORIES TBB::tbb INTERFACE_INCLUDE_DIRECTORIES)
# TBB_INTERFACE_INCLUDE_DIRECTORIES=/miniconda3/envs/custom-cgal/include

get_target_property(CGAL_INTERFACE_SYSTEM_INCLUDE_DIRECTORIES CGAL::CGAL INTERFACE_SYSTEM_INCLUDE_DIRECTORIES)
# CGAL_INTERFACE_SYSTEM_INCLUDE_DIRECTORIES=/miniconda3/envs/custom-cgal/include/;/miniconda3/envs/custom-cgal/include/;/miniconda3/envs/custom-cgal/include
get_target_property(CGAL_INTERFACE_INCLUDE_DIRECTORIES CGAL::CGAL INTERFACE_INCLUDE_DIRECTORIES)
# CGAL_INTERFACE_INCLUDE_DIRECTORIES=$<BUILD_INTERFACE:/workspace/contrib/CGAL-6.0-I-207/include>$<INSTALL_INTERFACE:include>/miniconda3/envs/custom-cgal/include;/miniconda3/envs/custom-cgal/include;/miniconda3/envs/custom-cgal/include
```
